### PR TITLE
C++: Fix `GuardCondition` comparison with constant pointers

### PR DESCRIPTION
**Pull request by MathiasVP**
Consider this (which I also added as test):
```cpp
void pointer_comparison(char* c) {
  if(c) {
    use(c);
  }
```
the IR for this essentially looks like:
```cpp
void pointer_comparison(char* c) {
  if(c != (char*)0) {
    use(c);
  }
```
All the logic is in place for ensuring that we know that `c != 0` inside the branch. However, we were overly restrictive when checking the type of the constant. The fix is very trivial, however.

### CODE REVIEW

**Comment by MathiasVP** [comment]
DCA is uneventful

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -1157,4 +1157,4 @@ private predicate add_eq(
 }
 
 /** The int value of integer constant expression. */
-private int int_value(Instruction i) { result = i.(IntegerConstantInstruction).getValue().toInt() }
+private int int_value(Instruction i) { result = i.(ConstantInstruction).getValue().toInt() }
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1160]: Wouldn't this mean that comparisons with the float `0.0` would now also be ok?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1160]: Hmm, good question. I would hope that `toInt` wouldn't have a result for that, but that's definitly worth a test. I guess we could place a restriction on the instruction saying that it should either be a `IntegerConstantInstruction` (which would cover `int` and `bool`) or a `PointerConstantInstruction` (or whatever we may call a constant instruction with a pointer type.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Well spotted! That was indeed a problem. I've force-pushed a new commit history to highlight the problem better more clearly, and also show that the fix actually prevents this problem.

**Comment by MathiasVP** [comment]
I'll run another DCA for good measure

**Pull request merged**